### PR TITLE
Add compression_codec field to system.part_columns table

### PR DESCRIPTION
**Pull request by ddelnano**
This change is to address #37311 and is includes part of the changes from the previous PR to address this (#37672). This change is not functional yet, but I need some feedback in how to best read the compression information from a part in order to handle in progress `ALTER`s to a column's codec (as discussed [here](https://github.com/ClickHouse/ClickHouse/pull/37672/files#r888146231)). 

The newly added test fails with the following output, so it seems that each part is labeled with the codec type used by the primary key. I believe that I need to read the codec info from the part's `.bin` file, but I haven't been able to figure out how each column's codec is specified in that file. I see that the primary key's codec is specified in the 3rd byte of that file (as documented [here](https://github.com/ClickHouse/ClickHouse/blob/4a192abe97b0c1da3a1fa08d45c10e493ea9c6ff/src/Compression/CompressionInfo.h#L20))

```
vagrant@ubuntu2204:/vagrant$ tests/clickhouse-test --no-drop-if-fail 02493_compression_codec_system_part_columns
WARNING: jinja2 not installed! Template tests will be skipped.
Using queries from '/vagrant/tests/queries' directory
Using clickhouse-client as client program (expecting split build)
Connecting to ClickHouse server... OK

Running 1 stateless tests (MainProcess).

02493_compression_codec_system_part_columns:                            [ FAIL ] - result differs with reference:
--- /vagrant/tests/queries/0_stateless/02493_compression_codec_system_part_columns.reference    2022-11-23 19:27:39.210808681 +0000
+++ /vagrant/tests/queries/0_stateless/02493_compression_codec_system_part_columns.stdout       2022-11-28 17:06:31.598497328 +0000
@@ -1,5 +1,5 @@
 id     CODEC(LZ4)
-data   CODEC(ZSTD(1))
-ddd    CODEC(NONE)
-somenum        CODEC(ZSTD(2))
-somestr        CODEC(LZ4HC(7))
+data   CODEC(LZ4)
+ddd    CODEC(LZ4)
+somenum        CODEC(LZ4)
+somestr        CODEC(LZ4)
````

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
- Add compression_codec field to system.part_columns table

### CODE REVIEW

File: src/Storages/MergeTree/IMergeTreeDataPart.h [hunk 1]
```diff
@@ -446,6 +446,10 @@ class IMergeTreeDataPart : public std::enable_shared_from_this<IMergeTreeDataPar
     void removeDeleteOnDestroyMarker();
     void removeVersionMetadata();
 
+    /// Found column without specific compression and return codec
+    /// for this column with default parameters.
+    CompressionCodecPtr detectDefaultCompressionCodec() const;
```
Comment (ddelnano) [inline src/Storages/MergeTree/IMergeTreeDataPart.h:451]: This change was made to test if this function would be able to determine each column's codec type from the `.bin` file on disk. It appears to only return the primary key's codec type, so it doesn't accomplish what we need.

I would appreciate any pointers if there is existing code that can determine a given part and column's codec type from a `.bin` file. I haven't been able to find that code myself and it seems like something that should exist.

**Comment by ddelnano** [comment]
@alesapin I'm trying to revive #37672, however, I am seeking some input for how to address this issue you had with the previous PR. Please see my PR description for more details.

**Comment by alesapin** [comment]
Actually this task is more complex than expected. I think we need to do the following:
### For wide parts
Implement a virtual method inside `IMergeTreeDataPart` similar to `detectDefaultCompressionCodec()` which takes column name as an argument. Logic is the following:
1. If column exists on disk (exists in columns.txt) -- just get any file of this column (similar to the logic of `detectDefaultCompressionCodec()`)  and read compression codec with `getCompressionCodecForFile`.
2. If column doesn't exists on disk == check column metadata StorageInMemoryMetadata. If some codec is specified there -- return it.
3. Check `DEFAULT_COMPRESSION_CODEC_FILE_NAME` -- if some code is specified, return it.
4. Return `CompressionCodecFactory::instance().getDefaultCodec()`
### For in-memory parts:
Always return `NONE`, even if something is specified in AST>
### For compact parts.
The most complex case. Similar to wide parts, but we will need to implement more low-level method than `getCompressionCodecForFile`. Because all compact parts columns are stored in a single file we first have to seek to the column offset (first one) and after that read the compression codec. All other steps are the same.

**Comment by ddelnano** [comment]
@alesapin thanks for that detailed explanation and that should be enough information for me to take another pass at this :+1:

**Comment by ddelnano** [comment]
@alesapin my latest commit is still a work in progress, but I have more questions.

For wide parts, I believe I'm close to a functional solution. Each of the columns is returning the codec specified on disk, however, I noticed that my test is seeing differences in the "level" portion of the compression codec. See the output below to see what I mean.
```
vagrant@ubuntu2204:/vagrant$ ./tests/clickhouse-test --no-drop-if-fail 02493_compression_codec_system_part_columns_wide

[ ... ]

+++ /vagrant/tests/queries/0_stateless/02493_compression_codec_system_part_columns_wide.stdout  2022-12-10 02:47:24.341980906 +0000
@@ -1,5 +1,5 @@
 id     CODEC(LZ4)
 data   CODEC(ZSTD(1))
 ddd    CODEC(NONE)
-somenum        CODEC(ZSTD(2))
-somestr        CODEC(LZ4HC(7))
+somenum        CODEC(ZSTD(1))
+somestr        CODEC(LZ4)

```

How is that level specified? That seems like something that would be internal to the data format of the given codec and I'm surprised that there is a difference. Also it seems there is an edge case handling LZ4 and LZ4HC. I know in the clickhouse file format they are the same, but I'm not sure if there is an actual difference there.

Once that is addressed and the code is cleaned up, the wide format should be in a decent place.

As for the compact parts, I'm struggling to determine what existing code is available to aid in computing the offset needed to seek for a given column. Can you advise what existing constructs can help there?

File: src/Storages/MergeTree/IMergeTreeDataPart.cpp [hunk 2]
```diff
@@ -986,8 +986,7 @@ CompressionCodecPtr IMergeTreeDataPart::detectDefaultCompressionCodec() const
     for (const auto & part_column : columns)
     {
         /// It was compressed with default codec and it's not empty
-        auto column_size = getColumnSize(part_column.name);
-        if (column_size.data_compressed != 0 && !storage_columns.hasCompressionCodec(part_column.name))
+        if (storage_columns.hasCompressionCodec(part_column.name))
```
Comment (ddelnano) [inline src/Storages/MergeTree/IMergeTreeDataPart.cpp:989]: This was from some debugging I did and is unrelated to my change. I noticed that the previous if statement was never true and so the files never get read from disk in `detectDefaultCompressionCodec`

**Comment by ddelnano** [comment]
@alesapin would appreciate if you could give this a look. I'd like to continue working on this, but want to make sure I'm on the right track.

**Pull request closed without merging**